src/altair/block_processing.rs

```diff
@@ -29,9 +46,123 @@ pub fn process_attestation<
+                upper_bound: data.slot + context.min_attestation_inclusion_delay,
+            },
+        )));
+    }
+
+    let committee_count = get_committee_count_per_slot(state, data.target.epoch, context);
+    if data.index >= committee_count {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::InvalidIndex {
+                index: data.index,
+                upper_bound: committee_count,
+            },
+        )));
+    }
+
+    let committee = get_beacon_committee(state, data.slot, data.index, context)?;
+    if attestation.aggregation_bits.len() != committee.len() {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::Bitfield {
+                expected_length: committee.len(),
+                length: attestation.aggregation_bits.len(),
+            },
+        )));
+    }
+
+    // Participation flag indices
+    let inclusion_delay = state.slot - data.slot;
+    let participation_flag_indices =
+        get_attestation_participation_flag_indices(state, data, inclusion_delay, context)?;
+
+    // Verify signature
+    let _ = is_valid_indexed_attestation(
+        state,
+        &mut get_indexed_attestation(state, attestation, context)?,
+        context,
+    )?;
+
+    // Update epoch participation flags
+    // @dev deviate from the order of the spec to avoid immutable borrow after mutable borrow
+    let attesting_indices =
+        get_attesting_indices(state, data, &attestation.aggregation_bits, context)?;
+
+    let epoch_participation = if is_current {
+        &mut state.current_epoch_participation
+    } else {
+        &mut state.previous_epoch_participation
+    };
+
+    let mut proposer_reward_numerator = 0;
+    // @dev this is where I start having trouble
+    // the `get_base_reward` below borrows state as immutable, but `epoch_participation` above borrows as mutable
+    for index in attesting_indices {
+        for (flag_index, weight) in crate::altair::PARTICIPATION_FLAG_WEIGHTS.iter().enumerate() {
+            // if flag_index in participation_flag_indices and not has_flag(epoch_participation[index], flag_index):
+            if participation_flag_indices.contains(&flag_index)
+                && !has_flag(epoch_participation[index], flag_index as u8)
+            {
+                epoch_participation[index] = add_flag(epoch_participation[index], flag_index as u8);
+                // @dev explicit import of `get_base_reward` to disambiguate instead of `spec` import
+                // also, this is where the issue with mutable vs immutable borrow occurs
+                /*
+                proposer_reward_numerator +=
+                    crate::altair::helpers::get_base_reward(state, index, context)? * weight;
+                */
+            }
+        }
+    }
+
+    // Reward proposer
+    let proposer_reward_denominator = (crate::altair::WEIGHT_DENOMINATOR
```

Comment:
cleaner to do this imo
```suggestion
    let proposer_reward_denominator = (WEIGHT_DENOMINATOR
```

and just pull all of these constants in from the `use spec::{..` up above

Reply:
Ahh, okay, yeah all of these were based on this comment, which I misinterpreted:
https://github.com/ralexstokes/ethereum-consensus/blob/65fc02162a59d70106f73f6b6984167161cb1c06/src/altair/helpers.rs#L229

In the example above, the const is _passed to a fn_ -- this is the only time where direct imports should be used for code gen simplification purposes. Otherwise, like you said, just do a standard top-level import.

Reply:
Although, another fn uses the direct imports mentioned here:
https://github.com/ralexstokes/ethereum-consensus/blob/65fc02162a59d70106f73f6b6984167161cb1c06/src/altair/helpers.rs#L297

Is there a rule of thumb between using direct imports (like you've done) vs. top-level?

Reply:
tbh not sure why I did this other than taking a confusing shortcut...

I can patch this up, no reason these shouldn't flow forward to each fork